
# setting a project-specific library
Sys.unsetenv("R_LIBS_USER")
dir.create("RLibrary")
.libPaths()
.libPaths(paste(getwd(), "RLibrary", sep="/"))
setRepositories()

# installing packages----
install.packages("beepr")
install.packages('BiocManager')
install.packages('tidyverse')
install.packages("tximport")
install.packages('ensembldb')
install.packages('rhdf5')
install.packages('EnsDb.Hsapiens.v86')
install.packages('datapasta')

BiocManager::install('tximport')
BiocManager::install('ensembldb')
BiocManager::install('rhdf5')
BiocManager::install('EnsDb.Hsapiens.v86')


# loading packages----
library(rhdf5) #provides functions for handling hdf5 file formats (kallisto outputs bootstraps in this format)
library(tidyverse) # provides access to Hadley Wickham's collection of R packages for data science, which we will use throughout the course
library(tximport) # package for getting Kallisto results into R
library(ensembldb) #helps deal with ensembl
library(EnsDb.Solyc.v86) # database package for tomato
library(readr)

# read data into R
targets <- read_tsv("~/course/Assign/trimmed/studydesign.txt")

# file paths to the abundance files generated by Kallisto using the 'file.path' function

path <- file.path(targets$sample, "abundance.tsv") 

# checking if this path is correct by seeing if the files exist
file.exists(path)

# annotations using organism-specific package ----
Tx <- transcripts(EnsDb.Solyc.v86, columns=c("tx_id", "gene_name"))
Tx <- as_tibble(Tx)
#need to change first column name to 'target_id'
Tx <- dplyr::rename(Tx, target_id = tx_id)
#transcrip ID needs to be the first column in the dataframe
Tx <- dplyr::select(Tx, "target_id", "gene_name")

# OPTIONAL: get annotations using BiomaRt----
# The annotation method described in the code chunk above works great if an organism-specific data base package exists for your organisms of interest
# however, this is only the case for human, mouse and rat....
# so, this optional code chunk shows one way you can get annotation data for other target organisms
# in this example, we're retrieving 1:1 mappings between transcript identifiers and gene symbols for tomato
library(biomaRt) # an alternative for annotation

listMarts() #default host is ensembl.org, and most current release of mammalian genomes
listMarts(host="parasite.wormbase.org") #access to parasite worm genomes
#listMarts(host="protists.ensembl.org") #access to protozoan genomes
listMarts(host="plants.ensembl.org") #access to plant genomes

#choose the 'mart' you want to work with
#myMart <- useMart(biomart="ENSEMBL_MART_ENSEMBL") # marts for mammalian genomes
myMart <- useMart(biomart="plants_mart", host="plants.ensembl.org") # marts for plant genomes

#take a look at all available datasets within the selected mart
available.datasets <- listDatasets(myMart)
#now grab the ensembl annotations for tomato
lyc.anno <- useMart(biomart="plants_mart", dataset = "slycopersicum_eg_gene", host="plants.ensembl.org")
lyc.attributes <- listAttributes(lyc.anno)
#take a look at all available attributes within the selected tomato annotation
Tx.lyc <- getBM(attributes=c('ensembl_transcript_id',
                             'ensembl_gene_id', 'description'),
                mart = lyc.anno)
#turn it into a table
Tx.lyc <- as_tibble(Tx.lyc) 
#we need to rename the two columns we just retreived from biomart
Tx.lyc <- dplyr::rename(Tx.lyc, target_id = ensembl_transcript_id, gene_name = ensembl_gene_id)
#transcrip ID needs to be the first column in the dataframe
Tx.lyc <- dplyr::select(Tx.lyc, "target_id", "gene_name")

Tx.lyc # mapping provided my biomart
#   target_id          gene_name       
#   <chr>              <chr>           
# 1 Solyc12g042120.2.1 Solyc12g042120.2
# 2 Solyc01g067380.3.1 Solyc01g067380.3
# 3 Solyc05g006260.1.1 Solyc05g006260.1
# 4 Solyc12g019340.1.1 Solyc12g019340.1
# 5 Solyc12g097070.2.1 Solyc12g097070.2
# 6 Solyc05g015710.3.1 Solyc05g015710.3
# 7 Solyc09g057715.1.1 Solyc09g057715.1
# 8 Solyc06g072480.1.1 Solyc06g072480.1
# 9 Solyc10g008210.3.1 Solyc10g008210.3
#10 Solyc12g009745.1.1 Solyc12g009745.1
# ... with 35,815 more rows

## optional if needed:
#Tx.lyc = apply(Tx.lyc,2, function(x) gsub("\\.[0-9]+$","", x))
#Tx.lyc <- as_tibble(Tx.lyc) 

#Tx.lyc # mapping truncated for isoforms
#   target_id        gene_name     
#   <chr>            <chr>         
# 1 Solyc12g042120.2 Solyc12g042120
# 2 Solyc01g067380.3 Solyc01g067380
# 3 Solyc05g006260.1 Solyc05g006260
# 4 Solyc12g019340.1 Solyc12g019340
# 5 Solyc12g097070.2 Solyc12g097070
# 6 Solyc05g015710.3 Solyc05g015710
# 7 Solyc09g057715.1 Solyc09g057715
# 8 Solyc06g072480.1 Solyc06g072480
# 9 Solyc10g008210.3 Solyc10g008210
#10 Solyc12g009745.1 Solyc12g009745
# ... with 35,815 more rows

# import Kallisto transcript counts into R using Tximport ----
# copy the abundance files to the working directory and rename so that each sample has a unique name
Txi_gene <- tximport(path, 
                     type = "kallisto", 
                     tx2gene = Tx, #Mapping from transcript IDs to gene IDs 
                     txOut = T, #How does the result change if this =FALSE vs =TRUE?
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)

Txi_trans <- tximport(path, 
                     type = "kallisto", 
                     tx2gene = Tx, #Mapping from transcript IDs to gene IDs 
                     txOut = TRUE, #How does the result change if this =FALSE vs =TRUE?
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)


#take a look at the type of object you just created
class(Txi_gene)
names(Txi_gene)



# the essentials ----
# this chunk contains the minimal essential code from this script. Simply uncomment the lines below and run the code.
library(tidyverse) # provides access to Hadley Wickham's collection of R packages for data science, which we will use throughout the course
library(tximport) # package for getting Kallisto results into R
library(ensembldb) #helps deal with ensembl
library(EnsDb.Hsapiens.v86) #replace with your organism-specific database package
targets <- read_tsv("~/course/Assign/trimmed/studydesign.txt")# read in your study design
path <- file.path(targets$sample, "abundance.tsv") # set file paths to your mapped data
Tx <- transcripts(EnsDb.Hsapiens.v86, columns=c("tx_id", "gene_name"))
Tx <- as_tibble(Tx)
Tx <- dplyr::rename(Tx, target_id = tx_id)
Tx <- dplyr::select(Tx, "target_id", "gene_name")
Txi_gene <- tximport(path, 
                     type = "kallisto", 
                     tx2gene = Tx, 
                     txOut = T, #determines whether your data represented at transcript or gene level
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)

